File: src/main/java/com/dozycoffee/wms/global/common/SoftDeletableEntity.java
package com.dozycoffee.wms.global.common;

import lombok.Getter;

import java.time.LocalDateTime;

@Getter
public abstract class SoftDeletableEntity extends BaseEntity {

    private LocalDateTime deletedAt;
    private String deletedBy;

    public boolean isDeleted() {
        return deletedAt != null;
    }

    protected void softDelete(String actor) {
        this.deletedAt = LocalDateTime.now();
        this.deletedBy = actor;
    }

    @Override
    public void copyAuditFieldsFrom(BaseEntity existing) {
        super.copyAuditFieldsFrom(existing);
        if (existing instanceof SoftDeletableEntity softDeletableEntity) {
            this.deletedAt = softDeletableEntity.deletedAt;
            this.deletedBy = softDeletableEntity.deletedBy;
        }
    }
}

File: src/main/kotlin/com/dozycoffee/wms/global/common/SoftDeletableEntity.kt
package com.dozycoffee.wms.global.common

import java.time.LocalDateTime

abstract class SoftDeletableEntity : BaseEntity() {

    var deletedAt: LocalDateTime? = null
        private set

    var deletedBy: String? = null
        private set

    fun isDeleted(): Boolean = deletedAt != null

    protected fun softDelete(actor: String) {
        this.deletedAt = LocalDateTime.now()
        this.deletedBy = actor
    }
}
